[MLIR] Vector: turn the ExtractStridedSlice rewrite pattern from #111541 into a canonicalization

bjacob opened the pull request:
This is a reasonable canonicalization because `extract` is more constrained than `extract_strided_slices`, so there is no loss of semantics here, just lifting an op to a special-case higher/constrained op. And the additional `shape_cast` is merely adding leading unit dims to match the original result type.

Context: discussion on #111541. I wasn't sure how this would turn out, but in the process of writing this PR, I discovered at least 2 bugs in the pattern introduced in #111541, which shows the value of shared canonicalization patterns which are exercised on a high number of testcases.

bjacob committed:
canonicalize

joker-eph commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -3772,6 +3772,92 @@ class StridedSliceSplat final : public OpRewritePattern<ExtractStridedSliceOp> {
+    for (int i = 0; i < numOffsets; ++i) {
+      if (sizes[i] != 1) {
+        return failure();
+      }
```
Nit: extra braces on this if (and the 3 others above)

joker-eph commented on mlir/lib/Dialect/Vector/IR/VectorOps.cpp:
```diff
@@ -3772,6 +3772,92 @@ class StridedSliceSplat final : public OpRewritePattern<ExtractStridedSliceOp> {
+///         %0 = vector.extract %arg0[0, 0, 0, 0] : vector<8xi8> from
+///         vector<8x1x1x2x8xi8> %1 = vector.shape_cast %0 : vector<8xi8> to
+///         vector<1x1x1x1x8xi8>
+///
```
Maybe document the casts insertion as well?

bjacob replied:
That was clang-format eating the newline and making it look like a single op.

joker-eph reviewed (approved):
Looks good to me, but just please update the description to justify why it is a good canonicalization and state that we don't see any loss of semantics in the transformation here.

Thanks!

bjacob committed:
review comments and fix

Signed-off-by: Benoit Jacob <jacob.benoit.1@gmail.com>

bjacob commented:
The 2nd commit pushed here addresses review comments AND fixes a bug (thankfully just by deleting a few lines).  I was over-zealously trying to prevent leading unit dims in the `extract` result type and in the process I was causing `numOffsets` to exceed `offsets.size()` which would cause an out-of-bounds read.

banach-space reviewed (approved):
Makes sense, thanks!
